src/Integration.Vsix/InfoBar/InfoBarManager.cs

```diff
@@ -0,0 +1,245 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="InfoBarManager.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using Microsoft.VisualStudio;
+using Microsoft.VisualStudio.Imaging;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using SonarLint.VisualStudio.Integration.InfoBar;
+using SonarLint.VisualStudio.Integration.Vsix.Resources;
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Globalization;
+
+namespace SonarLint.VisualStudio.Integration.Vsix.InfoBar
+{
+    [Export(typeof(IInfoBarManager))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    internal class InfoBarManager : IInfoBarManager
+    {
+        private readonly IServiceProvider serviceProvider;
+
+        [ImportingConstructor]
+        public InfoBarManager([Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            this.serviceProvider = serviceProvider;
+        }
+
+        #region IInfoBarManager
+        public IInfoBar AttachInfoBar(Guid toolwindowGuid, string message, string buttonText, ImageMoniker imageMoniker)
+        {
+            if (string.IsNullOrWhiteSpace(message))
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+
+            if (string.IsNullOrWhiteSpace(buttonText))
+            {
+                throw new ArgumentNullException(nameof(buttonText));
+            }
+
+            IVsWindowFrame frame = GetToolWindowFrame(this.serviceProvider, toolwindowGuid);
+
+            InfoBarModel model = CreateModel(message, buttonText, imageMoniker);
+
+            IVsInfoBarUIElement uiElement = null;
+            if (TryCreateInfoBarUI(this.serviceProvider, model, out uiElement)
+                && TryAddInfoBarToFrame(frame, uiElement))
+            {
+                return new PivateInfoBarWrapper(frame, uiElement);
+            }
+
+            return null;
+        }
+
+        public void DetachInfoBar(IInfoBar currentInfoBar)
+        {
+            if (currentInfoBar == null)
+            {
+                throw new ArgumentNullException(nameof(currentInfoBar));
+            }
+
+            PivateInfoBarWrapper wrapper = currentInfoBar as PivateInfoBarWrapper;
```

Comment:
Issue S3215 here regarding the method supposedly accepting an `IInfoBar` but actually requiring a specific implementation.
Am I right in thinking that the rationale for having the public `IInfoBar` interface and the private `PrivateInfoBarWrapper` is because...

1. we only want consumers using the `InfoBarManager` to create the info bars, and
2. we return the created info bar (as `IInfoBar`) only so that they have handle in order to detach it later?